core/src/main/java/com/linecorp/armeria/common/multipart/MultipartDecoder.java

```diff
@@ -98,16 +120,236 @@ public long demand() {
+            delegatedSubscriberUpdater.compareAndSet(this, null, abortedSubscriber);
+        }
         decoded.abort(cause);
     }
+
+    BodyPartPublisher onBodyPartBegin() {
+        return new BodyPartPublisher();
+    }
+
+    void requestUpstreamForBodyPartData() {
+        @Nullable
+        final MultipartSubscriber delegatedSubscriber = this.delegatedSubscriber;
+        if (delegatedSubscriber != null) {
+            delegatedSubscriber.requestUpstreamForBodyPartData();
+        }
+    }
+
+    /**
+     * Called by delegated MultipartSubscriber.
+     * e.g.
+     * 1. MultipartDecoder#abort -> MultipartSubscriber#onError -> MultipartSubscriber#cleanup ->
+     *    MultipartDecoder#cleanup
+     * 2. MultipartSubscriber#onComplete -> MultipartSubscriber#cleanup -> MultipartDecoder#cleanup
+     */
+    private void cleanup() {
+        if (delegatedSubscriber != null) {
+            delegatedSubscriber = null;
+        }
+    }
+
+    final class BodyPartPublisher extends DefaultStreamMessage<HttpData> {
+        @Override
+        protected void onRequest(long n) {
+            // Because whenConsumed will run in the same thread(called by onRequest) after looping the existing
+            // queue.(onRequest & event notification in whenConsumed will run in the same executor specified
+            // at subscribe)
+            // So if there is no change, it means there is no buffered data processed.
+            whenConsumed().thenRun(() -> {
+                // There isn't any buffered data, or it's not enough
+                if (demand() > 0) {
+                    requestUpstreamForBodyPartData();
+                }
+            });
+        }
+    }
+
+    /**
+     * For hiding Subscriber interface from MultipartDecoder.
+     * Subscribing {@code DecodedHttpStreamMessage<BodyPart>} and delegating to external subscriber.
+     */
+    private final class MultipartSubscriber implements Subscriber<BodyPart>, Subscription {
+        private final Subscriber<? super BodyPart> subscriber;
+        private final EventExecutor executor;
+        @Nullable
+        private Subscription subscription;
+        @Nullable
+        private StreamMessage<? extends HttpData> currentExposedBodyPartPublisher;
+        private boolean cancelled;
+
+        private MultipartSubscriber(Subscriber<? super BodyPart> subscriber, EventExecutor executor) {
+            this.subscriber = subscriber;
+            this.executor = executor;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            this.subscription = subscription;
+            subscriber.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(BodyPart bodyPart) {
+            if (demandOfMultipart != Long.MAX_VALUE) {
+                demandOfMultipart--;
+            }
+
+            currentExposedBodyPartPublisher = bodyPart.content();
+            // We only publish one bodyPart one time.
```

Comment:
nit:
```suggestion
            // We only publish one bodyPart at a time.
```
How about just `currentBodyPart`?